### PR TITLE
feat(gcp): Phase 0 — GCP adapter + Cloud Storage emulator

### PR DESCRIPTION
**Pull request by aamerik**
## Description

Implements **Phase 0** of the GCP emulator: the `jaiscloud-gcp` binary now speaks
the Google Cloud Storage wire protocol end-to-end, replacing the previous 501
stub. This is the first phase of the multi-phase GCP effort.

### What's in this PR

- **`GCPAdapter`** (`internal/gcp/adapter/`) — full `CloudAdapter` implementation
  with URL-path-based service detection (replaces the 501 stub).
- **Identity** (`internal/gcp/identity/`) — unverified OAuth2 bearer-token JWT
  parsing (service-account email/sub) + project-ID-from-path.
- **Resource names** (`internal/gcp/resource/`) — GCP hierarchical name formatters
  (`projects/{project}/...`).
- **Cloud Storage provider** (`internal/gcp/provider/storage/`) on the shared
  `ResourceStore` + `BlobStore`:
  - buckets: insert/get/update/delete/list (paginated)
  - objects: upload (media, multipart, resumable multi-chunk with `308 Resume
    Incomplete` + offset repair), download (`?alt=media` + raw `/{bucket}/{object}`
    media path used by the Go SDK), list (paginated + `prefix` + `delimiter`,
    composable), **objects.update (PUT) strict-replace vs objects.patch (PATCH)
    merge**, delete
  - bucket + **object** IAM (`getIamPolicy`/`setIamPolicy` with etag-based
    optimistic concurrency control)
  - bucket + object ACL (`list`/`insert`)
  - resumable sessions: atomic monotonic IDs, 1000-session cap, 24h TTL sweep,
    spill-to-temp-file beyond 4 MiB; Go-SDK upload protocol (POST chunks,
    `X-GUploader-No-308`, `bytes */N` status-query finalization, absolute
    `Location`)
- **GCP metadata-server emulator** (`internal/gcp/adapter/metadata.go`) — mocked
  ADC token issuance, `Metadata-Flavor` enforcement.
- **Config** — `ProjectID`, `GCPServiceAccount`, `GCPMetadataEnabled`
  (`JAISCLOUD_*`), cloud-specific isolation (`config.Load(cloud)`).
- **Wiring** — `cmd/jaiscloud-gcp/main.go`: snapshot-loop persistence
  (`state.json`), `env`/`reset`/`export`/`import`/`snapshot` subcommands,
  time-control flags.

### Out of scope (later phases)

- Pub/Sub, KMS, Secret Manager, IAM (Phase 1), Firestore/Functions/Workflows
  (Phase 2), logging/monitoring/eventing (Phase 3).
- Objects are stored single-generation (an insert overwrites the prior object
  rather than retaining multiple generations).

## How to test

```bash
# build + unit tests
make build-gcp
go test ./internal/gcp/...

# integration tests (requires the GCP binary on :8080)
./jaiscloud-gcp start &
go test -race ./tests/integration/gcp/

# official Go SDK tests (separate module)
cd tests/integration/gcp/sdk
STORAGE_EMULATOR_HOST=http://localhost:8080 go test -race ./...

# Postgres persistence e2e (requires JAISCLOUD_DSN)
go test -tags gcp_persistence ./tests/persistent_mode/gcp/storage/
```

## Test report

Automated against a live `jaiscloud-gcp` pod (k3d, `fcf13ed`):

- Unit: 39 provider + 15 adapter tests (`-race`), provider coverage **91.5%**.
- Integration: 14 wire tests (`tests/integration/gcp/`) — acceptance flow,
  multipart, PUT/PATCH semantics, object IAM + 404, etag-OCC 409, delimiter +
  pagination, ACL, `%2F` names, 50-way concurrency, resumable multi-chunk.
- Official Go SDK (`cloud.google.com/go/storage` over `STORAGE_EMULATOR_HOST`):
  write/read/list/delete + 8 MiB chunked resumable round-trip
  (`tests/integration/gcp/sdk/`, CI job `test-gcp`).
- Postgres persistence across restart (`tests/persistent_mode/gcp/storage/`).

### CODE REVIEW

**Comment by aamerik** [comment]
@jaisrajms — Phase 0 (GCP adapter + Cloud Storage emulator) is ready for review. Commits squashed; test report is in the PR description.

**Comment by aamerik** [comment]
@jaisrajms — Phase 1 (Pub/Sub, Secret Manager, Cloud KMS, IAM) is ready and stacked on top of this PR: https://github.com/aamerik/jaiscloud/pull/1

It is a dependent PR (base = feat/gcp-phase0), so its diff shows only Phase 1 relative to this branch. Once this Phase 0 PR merges to main, I will retarget Phase 1 to jaisrajms/jaiscloud:main and it will show only the Phase 1 changes.

File: internal/config/config.go [hunk 1]
```diff
@@ -144,6 +155,9 @@ func Load() (*Config, error) {
 	viper.SetDefault("aws_emulator_endpoint", "")
 	viper.SetDefault("s3_virtual_host_bases", "")
 	viper.SetDefault("imds_enabled", false)
+	viper.SetDefault("gcp_project_id", "jaiscloud-project")
```
Comment (jaisrajms) [inline internal/config/config.go]: we should have a separate list of AWS specific configuration default like  s3_virtual_host_bases, and for gcp like gcp_service_account, and a common config default. Check which cloud mode jaiscloud is running and only set common and chosen cloud configuration, we should not support other cloud configuration. Follow this rule for all configuration, routing, resources etc.

File: internal/gcp/provider/storage/storage.go [hunk 2]
```diff
@@ -0,0 +1,872 @@
...
+
+// ─── objects ──────────────────────────────────────────────────────────────────
+
+func (p *Provider) ObjectsList(ctx context.Context, nr *model.NormalizedRequest) (*model.ProviderResponse, error) {
+	bucket, _ := nr.Params["bucket"].(string)
+	if bucket == "" {
+		return nil, model.NewProviderError("InvalidRequest", "missing bucket", 400)
+	}
+	prefix := bucket + "/"
+	entries, err := p.resources.List(ctx, nr.AccountID, store.GlobalRegion, rtObject, prefix)
```
Comment (jaisrajms) [inline internal/gcp/provider/storage/storage.go:340]: **Blocking — cross-bucket listing leak.** `p.resources.List(ctx, nr.AccountID, store.GlobalRegion, rtObject, bucket+"/")` relies on the store's `prefix` argument being a *true prefix*. It is not: `MemoryResourceStore.List` uses `strings.Contains(e.ID, prefix)` (`internal/store/memory.go:131`) and `PostgresResourceStore.List` uses `id LIKE '%' || $prefix || '%'` (`internal/store/postgres.go:184`). Because object IDs are `bucket/object`, any sibling bucket whose name contains `bucket/` as a substring (e.g. `abkt`, `mybkt`) leaks into this listing. The non-delimiter path's `filtered := entries[:0]` re-filter (line 306) does not reject these — when `pfx==""` every entry is kept, and `strings.TrimPrefix(e.ID, prefix)` is a no-op for foreign IDs. `BucketsDelete` (line 239) has the same bug and will spuriously return 409 for a non-empty check. Fix client-side with `strings.HasPrefix(strings.TrimPrefix(e.ID, bucket+"/"), ...)` on every path; do not change the shared store's `Contains` semantics (AWS and cross-scope scans depend on it).

File: internal/gcp/provider/storage/storage.go [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+	o.Md5Hash = base64.StdEncoding.EncodeToString(sum[:])
+	crc := crc32.Checksum(media, crc32.MakeTable(crc32.Castagnoli))
+	crcBytes := make([]byte, 4)
+	binary.BigEndian.PutUint32(crcBytes, crc)
+	o.Crc32c = base64.StdEncoding.EncodeToString(crcBytes)
+	o.ID = bucket + "/" + object + "/" + o.Generation
+	o.Etag = "CAE="
+	o.SelfLink = "https://www.googleapis.com/storage/v1/b/" + bucket + "/o/" + url.PathEscape(object)
+	o.MediaLink = "https://www.googleapis.com/download/storage/v1/b/" + bucket + "/o/" + url.PathEscape(object) + "?alt=media"
+	if err := p.blobs.Put(ctx, blobsNamespace, id, media); err != nil {
```
Comment (jaisrajms) [inline internal/gcp/provider/storage/storage.go]: **Blocking — full buffering, no streaming (high-scale performance).** `blobs.Put` buffers the entire object in memory, `ObjectsGetMedia` uses `blobs.Get` (line 429), and resumable uploads accumulate the whole object in `sess.buf` (line 820). Combined with the gateway's unconditional `io.ReadAll` for GCP (`isS3StreamingUpload` only matches S3 SigV4 paths in `gateway/server.go`), multi-GB media/resumable uploads and downloads will OOM the emulator. The AWS S3 provider streams: `_streaming` marker → read `r.Body` directly, `blobs.PutStream` (`object.go:287`), `blobs.GetStream(offset,length)` + `_stream` response (`crud.go:311,329`, `s3.go:708`). Mirror that for GCS: detect `uploadType=media|multipart|resumable` and `alt=media` download from headers/query before reading the body, stream uploads via `PutStream`, stream downloads via `GetStream` + `_stream`, and spill resumable chunks to a temp file instead of `sess.buf`.

File: internal/gcp/provider/storage/storage.go [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+	buf         []byte // accumulated bytes across chunks
+}
+
+// New returns a GCS provider backed by the given metadata and blob stores.
+func New(resources store.ResourceStore, blobs blobfs.BlobStore) *Provider {
+	return &Provider{
+		resources: resources,
+		blobs:     blobs,
+		uploads:   make(map[string]*uploadSession),
+		gen:       clock.Now().UnixNano(),
```
Comment (jaisrajms) [inline internal/gcp/provider/storage/storage.go:88]: **Generation counter is not persisted/monotonic across restarts.** `gen` seeds from `clock.Now().UnixNano()` at `New()` time; `nextGen()` (line 71) then increments in-memory. With `--dsn`, stored generations are strings; on restart the counter re-seeds and can collide or *regress* (a new object can get a generation lower than an existing one). Real GCS generations are strictly monotonic and durable. Seed from `max(stored generation)+1` (or persist the counter) before advertising `--dsn` parity.

File: internal/gcp/provider/storage/storage.go [hunk 5]
```diff
@@ -0,0 +1,872 @@
...
+	binary.BigEndian.PutUint32(crcBytes, crc)
+	o.Crc32c = base64.StdEncoding.EncodeToString(crcBytes)
+	o.ID = bucket + "/" + object + "/" + o.Generation
+	o.Etag = "CAE="
+	o.SelfLink = "https://www.googleapis.com/storage/v1/b/" + bucket + "/o/" + url.PathEscape(object)
+	o.MediaLink = "https://www.googleapis.com/download/storage/v1/b/" + bucket + "/o/" + url.PathEscape(object) + "?alt=media"
+	if err := p.blobs.Put(ctx, blobsNamespace, id, media); err != nil {
+		return nil, err
+	}
+
```
Comment (jaisrajms) [inline internal/gcp/provider/storage/storage.go:543]: **Insert ordering: blob written before metadata.** `blobs.Put` (line 398) succeeds, then `resources.Create` fails with a non-`AlreadyExists` error → the blob is orphaned (metadata gone, blob present). The S3 convention is metadata-then-blob for deletes and recheck-on-read for gets; insert should be atomic in the same spirit (e.g. create metadata first, or roll back the blob on create failure).

File: internal/gcp/provider/storage/storage.go [hunk 6]
```diff
@@ -0,0 +1,872 @@
...
+		if body, ok := nr.Params["body"].(map[string]any); ok {
+			object, _ = body["name"].(string)
+		}
+	}
+	if bucket == "" || object == "" {
+		return nil, model.NewProviderError("InvalidRequest", "missing bucket or object name", 400)
+	}
+	ct, _ := nr.Params[wire.ContentTypeKey].(string)
+
+	id := fmt.Sprintf("%d", clock.Now().UnixNano())
```
Comment (jaisrajms) [inline internal/gcp/provider/storage/storage.go]: **Unbounded resumable-upload sessions + collision-prone ID.** `uploads` grows without bound — a client can open unlimited sessions, each holding an in-memory `sess.buf` (DoS). There is no TTL, cap, or GC; the provider only implements `Reset` (line 83) with no `Shutdown`/worker lifecycle (compare the `workers`/`wg + cancel + Shutdown` pattern used by EMR/Lambda). Additionally, `fmt.Sprintf("%d", clock.Now().UnixNano())` for the session ID (and `nextGen`) can collide under concurrency — use an atomic counter or UUID. Add session expiry via a `clock.RealNow()` sweep and a max-session cap.

File: internal/gcp/adapter/gcs.go [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+	case 400:
+		return "INVALID_ARGUMENT"
+	case 401:
+		return "UNAUTHENTICATED"
+	case 403:
+		return "PERMISSION_DENIED"
+	case 404:
+		return "NOT_FOUND"
+	case 409:
+		return "ALREADY_EXISTS"
```
Comment (jaisrajms) [inline internal/gcp/adapter/gcs.go:406]: **409 maps to `ALREADY_EXISTS`.** The non-empty-bucket-delete path returns `model.NewProviderError("Conflict", "bucket is not empty", 409)`, but `gcpStatusString` translates 409 → `ALREADY_EXISTS`. Real GCS returns 409 with `reason` semantics closer to `FAILED_PRECONDITION` for this case. Consider a dedicated code (e.g. `FAILED_PRECONDITION`) for non-empty delete.

File: internal/gcp/adapter/gcs.go [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+		if len(vs) > 0 {
+			params[k] = vs[0]
+		}
+	}
+}
+
+// parseJSON decodes a JSON body into a map, or returns nil for empty/JSON bodies
+// that are not objects (e.g. media). Never returns an error — decode failures
+// are surfaced by the provider as validation errors.
+func parseJSON(body []byte) map[string]any {
```
Comment (jaisrajms) [inline internal/gcp/adapter/gcs.go]: **`parseJSON` swallows malformed JSON.** A syntactically invalid body returns `nil`, so the provider reports "missing bucket name" (400) instead of an invalid-JSON error. Surfacing a decode error here would give callers a more accurate error.

File: internal/gcp/adapter/metadata.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+		rt.Get("/project/project-id", func(w http.ResponseWriter, req *http.Request) {
+			writeMetadataText(w, cfg.ProjectID)
+		})
+		rt.Get("/project/numeric-project-id", func(w http.ResponseWriter, req *http.Request) {
+			writeMetadataText(w, "0")
+		})
+		rt.Get("/instance/service-accounts/default/email", func(w http.ResponseWriter, req *http.Request) {
+			writeMetadataText(w, cfg.ServiceAccount)
+		})
+		rt.Get("/instance/service-accounts/default/token", func(w http.ResponseWriter, req *http.Request) {
```
Comment (jaisrajms) [inline internal/gcp/adapter/metadata.go:39]: **Metadata-server emulator gaps.** (1) Real GCS metadata servers require the `Metadata-Flavor: Google` header (SSRF protection) — not enforced here. (2) `writeMetadataText` (line 52) does not set `Content-Length`, so responses are chunked. (3) This file has **no tests** — the token/mockAccessToken and route handlers should be covered, mirroring the AWS IMDS tests.

File: cmd/jaiscloud-gcp/main.go [hunk 10]
```diff
@@ -56,26 +100,57 @@ func startCmd() *cobra.Command {
...
+	viper.BindPFlag("ephemeral", cmd.Flags().Lookup("ephemeral"))
+	viper.BindPFlag("dsn", cmd.Flags().Lookup("dsn"))
+	viper.BindPFlag("data_dir", cmd.Flags().Lookup("data-dir"))
+	viper.BindPFlag("log_level", cmd.Flags().Lookup("log-level"))
+	viper.BindPFlag("metrics", cmd.Flags().Lookup("metrics"))
+	viper.BindPFlag("gcp_metadata_enabled", cmd.Flags().Lookup("gcp-metadata"))
+}
+
+func initStores(ctx context.Context, cfg *config.Config) (store.ResourceStore, blobfs.BlobStore, func(), error) {
+	if cfg.DSN != "" {
```
Comment (jaisrajms) [inline cmd/jaiscloud-gcp/main.go:237]: **`--ephemeral` + `--dsn` not mutually exclusive.** `initStores` checks `if cfg.DSN != ""` and silently prefers Postgres; the AWS binary/CLAUDE.md declares this a startup error. Add the mutual-exclusion check (and consider a `--fresh-start` flag for parity with AWS).

File: internal/gcp/resource/resource.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	"log/slog"
+)
+
+// formatters maps abstract resource types to their GCP resource-name format
+// function. To add a new resource type, add one entry here — no switch to update.
+var formatters = map[string]func(project, name string) string{
+	// Google Cloud Storage — bucket names are global, no project prefix.
+	"gcs-bucket": func(_, n string) string { return n },
+	"gcs-object": func(_, n string) string { return n },
+	// GCS bucket IAM policy resourceId uses a fixed "_" project placeholder.
```
Comment (jaisrajms) [inline internal/gcp/resource/resource.go:20]: **Forward-looking dead code.** The pubsub/kms/secret/service-account/cloud-function formatters (and the speculative `locOf`/`ringOf`/`keyOf` parsers) have no provider consumer yet. Harmless, but consider deferring these until Phase 1 to avoid unverified code paths.

**Review by jaisrajms — CHANGES_REQUESTED** [review]
## Review — GCP Phase 0 (adapter + Cloud Storage)

Overall this is a well-structured Phase 0: cloud isolation is respected (verified zero `internal/aws` imports under `internal/gcp/`), the `ServiceDescriptor`/`serviceProviderMap` table mirrors the AWS `services.go` pattern, and `nr.ResourceID(...)` is used for resource names. Two correctness/scale issues block merge; the rest are medium/minor and test-coverage parity.

### Blocking
1. **Cross-bucket listing/deletion leak** — the shared store's `List` `prefix` arg is a *substring* match (`strings.Contains` / `LIKE '%..%'`), not a prefix. Object IDs are `bucket/object`, so `bkt` sees `abkt`'s objects. See inline comments.
2. **No streaming** — media/resumable uploads and downloads are fully buffered in memory. The AWS S3 provider streams via `PutStream`/`GetStream`/`_stream`; GCS should too.

### Test coverage parity gap (vs AWS S3)
- Integration tests drive **raw HTTP**, not the official `cloud.google.com/go/storage` client. The PR body claims manual verification against the Go client, but there is no automated SDK-based test (`STORAGE_EMULATOR_HOST`).
- No `multipart` upload test (`parseMultipart` is untested), no metadata-server emulator tests, no negative/edge cases (409 already-exists, 404s, delimiter+prefix, pageToken edge cases), no concurrency tests for the generation counter / resumable-session map, and no postgres/persistent-mode e2e (AWS has `tests/persistent_mode/aws/`).

### What's done well
- Isolation: `internal/gcp/` imports only shared infra (`store`, `blobfs`, `model`, `provider`, `adapter`, `clock`, `admin`, `gateway`, `config`); no `internal/aws` import.
- Resource abstraction: `nr.ResourceID("gcs-bucket-policy", bucket)` + centralized formatters in `internal/gcp/resource`.
- `wire` package namespaces reserved keys (`jaiscloud:resumeLocation`, `jaiscloud:range`) to avoid collision with GCS resource fields.
- Clock discipline: `clock.Now()` for business timestamps, `clock.RealNow()` for token expiry.

See inline comments for the specific fixes.

**Comment by aamerik** [comment]
@jaisrajms — thanks for the thorough review. Addressed in `a5206da` (config isolation) and `d52b6e2` (correctness/scale/coverage):

**Config isolation** — `config.Load(cloud)` now registers/reads only the common keys plus the active cloud's keys; a GCP binary no longer sets AWS config (`s3_virtual_host_bases`, `imds`, `aws_emulator_endpoint`, k8s/spark/lambda, `kms`, `oidc`) and vice versa. Added `TestConfig_CloudSpecificIsolation`.

**Blocking — cross-bucket listing/deletion leak** — fixed client-side by filtering `List` results to the true `bucket/` prefix (both object listing and the non-empty-delete check). The shared store's substring semantics are unchanged.

**Blocking — streaming** — media uploads now stream via `PutStream` (md5/crc32c computed on the fly with a `TeeReader`, so size/checksums stay correct); downloads stream via `GetStream` + `_stream`. Large media objects no longer buffer fully in memory.

**Other inline fixes:**
- non-empty bucket delete → reason `bucketNotEmpty` (was `ALREADY_EXISTS`)
- insert rolls back the blob if the metadata create fails (no orphaned bytes)
- generation counter seeds from the max stored generation (monotonic across `--dsn` restarts)
- resumable sessions capped at 1000, and use the atomic counter for collision-free IDs
- `parseJSON` surfaces malformed JSON as a 400 instead of a misleading "missing field" error
- metadata-server enforces `Metadata-Flavor: Google` + sets `Content-Length`; added route + token tests
- `--ephemeral` + `--dsn` mutual exclusion is enforced in the shared `config.Load` (startup error)

**Acknowledged, deferred:**
- resumable/multipart uploads still buffer (`sess.buf`), not spilled to a temp file — that's the remaining streaming piece.
- SDK-based automated test (`cloud.google.com/go/storage` over `STORAGE_EMULATOR_HOST`) + multipart/negative/concurrency/persistent-mode e2e — the test-coverage-parity tranche.
- `--fresh-start` flag for AWS parity.

Re `resource.go` formatters: the pubsub/kms/secret/service-account/cloud-function formatters are consumed by Phase 1 (stacked in PR #1), so they aren't dead once that lands.

**Comment by aamerik** [comment]
@jaisrajms — closing out the review in `fcf13ed` (wire-protocol parity + coverage):

**Blockers**
- Cross-bucket listing/deletion leak — fixed client-side with a true `bucket/` prefix filter (`d52b6e2`).
- Streaming — media uploads/downloads stream via `PutStream`/`GetStream`/`_stream`; resumable sessions spill to a temp file past 4 MiB (`d52b6e2`, `fcf13ed`).

**Inline comments**
- Config isolation (`config.Load(cloud)`) — `a5206da`.
- Generation counter seeded from max stored generation — `d52b6e2`, covered by `TestGenerationSeededAcrossRestart`.
- Insert rollback on metadata-create failure — `d52b6e2`.
- Resumable sessions: atomic monotonic IDs + 1000-session cap + 24h TTL sweep with spill cleanup — `d52b6e2` + `fcf13ed`, covered by `TestResumableSessionEdgeCases`, `TestResumableSessionTTLSweep`, `TestResetCleansSpilledSessions`.
- 409 non-empty-delete now maps to reason `bucketNotEmpty` — `d52b6e2`.
- `parseJSON` now surfaces malformed JSON as a 400 — `d52b6e2`.
- Metadata server: `Metadata-Flavor: Google` enforcement + `Content-Length` + unit tests — `d52b6e2`.
- `--ephemeral`/`--dsn` mutual exclusion lives in shared `config.Load` (`a5206da`); `--fresh-start` flag added (`fcf13ed`).
- Forward-looking formatters — now consumed by Phase 1 (PR #1).

**Coverage parity**
- SDK-based integration tests (`tests/integration/gcp/sdk`, `cloud.google.com/go/storage` over `STORAGE_EMULATOR_HOST`, incl. chunked resumable + raw-media read path).
- Postgres persistence e2e (`tests/persistent_mode/gcp/storage`, `gcp_persistence` tag).
- Multipart, delimiter+pagination, `%2F` names, concurrency, etag-OCC, object IAM/ACL, 404/409 negatives — unit + wire.
- Provider unit coverage: 91.5%.

Also fixed while closing the SDK gap: objects.update (PUT) strict-replace vs objects.patch (PATCH) merge semantics, objects.getIamPolicy/setIamPolicy, and SDK upload-protocol details (POST chunks, X-GUploader-No-308, bytes */N finalization, absolute Location).

All 11 review threads are resolved.

**Comment by aamerik** [comment]
@jaisrajms — FYI on CI: `test-gcp` and `docker-build` are green. The `test` job (AWS) is red on every run of this PR because the AWS integration package hits its 120s `-timeout` in `TestSQS_QueueDeletedRecentlyExpires` (61s sleep; `tests/integration/phase_a2a3_test.go:570`) — unrelated to this PR (no `internal/aws` or `tests/integration` AWS files are touched). You may want to bump that job's timeout separately.
